feat: schema-first audit events

aramissennyeydd opened the pull request:
## Hey, I just made a Pull Request!

Bit of an experiment to try and simplify how audit events are defined and keep the existing code clean instead of muddying it with either extra middleware at the route level or try/catch blocks in the application code itself.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

aramissennyeydd committed:
feat: schema-first audit events

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

schultzp2020 commented on packages/backend-openapi-utils/src/middlewares/auditor.ts:
```diff
@@ -0,0 +1,220 @@
+  };
+
+  return { success, error };
+}
```
```suggestion
}

```
EOF

schultzp2020 commented on packages/backend-openapi-utils/src/middlewares/auditor.ts:
```diff
@@ -0,0 +1,220 @@
+          .then(auditorEvent => {
+            // Only mark as success for 2xx responses
+            // For errors, the handler should catch and call auditorEvent.fail()
+            if (res.statusCode >= 200 && res.statusCode < 300) {
```
I believe this if statement is redundant because you check the status code on line 169.

aramissennyeydd committed:
more guarantees around async-ness and add support for captureResponse

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix type errors

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
move to dependency object and update log statements

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
non blocking

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
typo

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
add debug log

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix api report

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
Update packages/backend-openapi-utils/src/middlewares/auditor.ts

Co-authored-by: Paul Schultz <pschultz@pobox.com>
Signed-off-by: Aramis Sennyey <159921952+aramissennyeydd@users.noreply.github.com>

Rugvip commented on packages/backend-openapi-utils/report.api.md:
```diff
@@ -921,4 +945,22 @@ type UnknownIfNever<P> = [P] extends [never] ? unknown : P;
+// @public (undocumented)
+export interface WithAuditorEvent {
+  // (undocumented)
+  [AUDITOR_SYMBOL]?: Awaited<ReturnType<AuditorService['createEvent']>>;
```
My reaction was the opposite, do we really need to export `WithAuditorEvent`? 😅

schultzp2020 replied:
> is it safe to call `.success` or `.error` twice on AuditorService?

@aramissennyeydd yes, it should be.

Rugvip commented on plugins/catalog-backend/src/schema/openapi.yaml:
```diff
@@ -1117,6 +1163,14 @@ paths:
+        severityLevel: medium
+        meta:
+          actionType: create
+          captureFromRequest:
```
Perhaps a bit cleaner to move this one up a level and call it something like `captureMetaFromRequest`?

Rugvip reviewed (commented):
I like! 👍 

Pretty nice cleanup of plugin routers, and it's also always possible to fall back to manual events if needed. Also not too much new API surface either given that we can mostly rely on the existing auditor API

aramissennyeydd commented on plugins/catalog-backend/src/schema/openapi/generated/router.ts:
```diff
@@ -794,6 +794,16 @@ export const spec = {
         operationId: 'RefreshEntity',
         tags: ['Entity'],
         description: 'Refresh the entity related to entityRef.',
+        'x-backstage-auditor': {
```
annoyingly enough, this is part of the spec itself, https://swagger.io/specification/ -
```
Field Pattern | Type | Description
-- | -- | --
^x- | Any | Allows extensions to the OpenAPI Schema. The field name MUST begin with x-, for example, x-internal-id. Field names beginning x-oai- and x-oas- are reserved for uses defined by the OpenAPI Initiative. The value can be any valid JSON value (null, a primitive, an array, or an object.)

Specification Extensions
While the OpenAPI Specification tries to accommodate most use cases, additional data can be added to extend the specification at certain points.

The extensions properties are implemented as patterned fields that are always prefixed by x-.

Field Pattern	Type	Description
^x-	Any	Allows extensions to the OpenAPI Schema. The field name MUST begin with x-, for example, x-internal-id. Field names beginning x-oai- and x-oas- are reserved for uses defined by the [OpenAPI Initiative](https://www.openapis.org/). The value can be any valid JSON value (null, a primitive, an array, or an object.)
```

aramissennyeydd committed:
address PR feedback, update types and move capture params to top level

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

schultzp2020 commented on plugins/catalog-backend/src/schema/openapi/generated/router.ts:
```diff
@@ -1448,6 +1552,15 @@ export const spec = {
+          meta: {
+            queryType: 'by-entity',
+            captureFromRequest: {
+              params: ['kind', 'namespace', 'name'],
```
@Rugvip there are a few other breaking changes lingering around as well. For instance, dry runs are always considered medium whereas before it was considered low because the severity is no longer dynamic. I think the trade off is probably worth it for simplicity.

schultzp2020 commented on plugins/catalog-backend/src/schema/openapi/generated/router.ts:
```diff
@@ -1264,6 +1332,17 @@ export const spec = {
+          eventId: 'location-mutate',
+          severityLevel: 'medium',
+          meta: {
+            actionType: 'create',
```
> @schultzp2020 wdyt? These should just be copy-pasted from what we had before. Maybe also related to [#32148 (comment)](https://github.com/backstage/backstage/pull/32148#discussion_r2673005445) - do we even need this property or should this be communicated in the event id?

I considered using a single generic field like `variant` or `type`, but I think keeping `queryType` and `actionType` separate provides useful semantic value in audit logs. When reviewing events, you can immediately tell if an operation was a read (`queryType: by-uid`) vs a write (`actionType: delete`) without needing to cross reference the `eventId`. That said, I'm not strongly opposed to consolidating to a single field if consistency is preferred - the tradeoff is just losing the implicit read/write signal.

aramissennyeydd replied:
Could we add HTTP verbs and route here instead? With the ability to override, thinking something like
```
GET -> query
POST -> create
DELETE -> delete
PATCH -> modify
```

aramissennyeydd committed:
add new variant property

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix ancestry

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
revert config change

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
use pattern resolver to allow more dynamic metadata

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
do not expose pattern resolver to public api

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix entityRef template

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
support both request and response resolvers but not at once

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix copilot issues

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix test issues and allow primitive values to be retained

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

freben commented on plugins/catalog-backend/src/schema/openapi.yaml:
```diff
@@ -936,6 +953,10 @@ paths:
+      x-backstage-auditor:
+        eventId: entity-fetch
+        meta:
+          entityRef: '{{ request.params.kind }}/{{ request.params.namespace }}:{{ request.params.name }}'
```
There's also the matter of them needing to be lowercased in the `en-US` locale. The way it is made in here is risky.

aramissennyeydd committed:
Add auditor annotations for routes added during rebase

Master added POST /entities/by-query, POST /entity-facets,
POST /locations/by-query, and PUT /locations/:id using the manual
auditor.createEvent pattern while this branch was rebasing. Their
manual auditing was stripped in createRouter.ts to match the rest of
the schema-first approach; this adds the corresponding
x-backstage-auditor annotations and regenerates the OpenAPI server
code so those routes keep emitting audit events via the middleware.

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
Add regression tests for catalog auditor events, fix validate-entity audit gap

Add a comprehensive test suite asserting that every catalog-backend
route emits the expected audit event via the schema-first auditor
middleware, covering both success and failure paths. This caught a
real gap: POST /validate-entity's x-backstage-auditor annotation never
fired because validatorOptions.ignorePaths excludes that route from
express-openapi-validator's managed routes, so req.openapi is never
populated. Restore manual auditor.createEvent/success/fail calls for
that one route, removing the dead annotation. While restoring the
manual auditing, also fix a latent bug where the handler fell through
to a 200 response after already sending a 400 on validation failure.

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
Address Copilot review feedback on auditor middleware

Fixes an aborted-connection hang in the response-finish wait, an
inaccurate error message for NaN pattern values, and documents the
raw (non-canonicalized) entityRef casing in audit meta patterns.

Also removes the auditorMiddlewareFactory export in favor of auditor
and logger options on createValidatedOpenApiRouter and
createValidatedOpenApiRouterFromGeneratedEndpointMap, which wire up
auditing automatically. catalog-backend no longer manually applies
the success/error middleware.

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
Address latest Copilot review feedback on auditor middleware

Meta patterns without spaces around placeholders (e.g. {{request.params.id}})
were misclassified and silently dropped instead of being resolved or
rejected. Also corrects stub.ts docs claiming only OpenAPI 3.1 is supported,
clarifies the NaN error message, and adds changesets for the catalog-backend
and repo-tools changes that were missing them.

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
fix prettier

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>
